contracts/TapETH.sol

```diff
@@ -0,0 +1,542 @@
+     * the total supply of TapETH by the staking rewards and the swap fee.
+     */
+    function setTotalSupply(uint256 _amount) external {
+        require(pools[msg.sender], "TapETH: no pool");
+        require(_amount != 0, "TapETH: no pool");
+        uint256 _deltaBuffer = Math.min(buffer, _amount);
+        _totalSupply += _amount - _deltaBuffer;
+        totalRewards += _amount;
+        buffer -= _deltaBuffer;
+    }
+
+    /**
+     * @return the amount of shares owned by `_account`.
+     */
+    function sharesOf(address _account) external view returns (uint256) {
+        return _sharesOf(_account);
+    }
+
+    /**
+     * @return the amount of shares that corresponds to `_tapETHAmount` protocol-controlled tapETH.
+     */
+    function getSharesByPooledEth(
+        uint256 _tapETHAmount
+    ) public view returns (uint256) {
+        return (_tapETHAmount * _getTotalShares()) / _getTotalPooledEther();
+    }
+
+    /**
+     * @return the amount of tapETH that corresponds to `_sharesAmount` token shares.
+     */
+    function getPooledEthByShares(
+        uint256 _sharesAmount
+    ) public view returns (uint256) {
+        if (totalShares == 0) {
+            return 0;
+        } else {
+            return (_sharesAmount * _getTotalPooledEther()) / (totalShares);
+        }
+    }
+
+    /**
+     * @notice Moves `_sharesAmount` token shares from the caller's account to the `_recipient` account.
+     * @return amount of transferred tokens.
+     * Emits a `TransferShares` event.
+     * Emits a `Transfer` event.
+     * @dev The `_sharesAmount` argument is the amount of shares, not tokens.
+     */
+    function transferShares(
+        address _recipient,
+        uint256 _sharesAmount
+    ) external returns (uint256) {
+        _transferShares(msg.sender, _recipient, _sharesAmount);
+        uint256 tokensAmount = getPooledEthByShares(_sharesAmount);
+        _emitTransferEvents(
+            msg.sender,
+            _recipient,
+            tokensAmount,
+            _sharesAmount
+        );
+        return tokensAmount;
+    }
+
+    /**
+     * @notice Moves `_sharesAmount` token shares from the `_sender` account to the `_recipient` account.
+     *
+     * @return amount of transferred tokens.
+     * Emits a `TransferShares` event.
+     * Emits a `Transfer` event.
+     *
+     * Requirements:
+     * - the caller must have allowance for `_sender`'s tokens of at least `getPooledEthByShares(_sharesAmount)`.
+     * @dev The `_sharesAmount` argument is the amount of shares, not tokens.
+     */
+    function transferSharesFrom(
+        address _sender,
+        address _recipient,
+        uint256 _sharesAmount
+    ) external returns (uint256) {
+        uint256 tokensAmount = getPooledEthByShares(_sharesAmount);
+        _spendAllowance(_sender, msg.sender, tokensAmount);
+        _transferShares(_sender, _recipient, _sharesAmount);
+        _emitTransferEvents(_sender, _recipient, tokensAmount, _sharesAmount);
```

Comment:
```suggestion
        _spendAllowance(_sender, msg.sender, _sharesAmount);
        _transferShares(_sender, _recipient, _sharesAmount);
        uint256 tokensAmount = getPooledEthByShares(_sharesAmount);
        _emitTransferEvents(_sender, _recipient, tokensAmount, _sharesAmount);
```

Reply:
- spendAllowance should be applied on tokensAmount and not shareAmount

Reply:
I agree. so please change the `shareAmount ` to `tokensAmount ` of the `_spendAllowance`  in `burnSharesFrom` .

Reply:
ok, was done ,  in the function "burnSharesFrom",  I used shareAmount in  "_spendAllowance" due to the  burn design explained this morning: the user burns only his shares and not his collected rewards and fee.